Cj4 docs fixes (#253)
leaningtech/labs: sites/cheerpj/src/content/docs/21-migrating-from-cheerpj2.md

title: Migration from CheerpJ 2
---

> CheerpJ version 3 (and above) is a complete reimplementation of CheerpJ 2, and as such it is not fully backward compatible. This page shows how to migrate from CheerpJ 2 to CheerpJ 3 (and above), which is applicable for CheerpJ 4.

Reviewer: ```suggestion
CheerpJ version 3 (and above) is a complete reimplementation of CheerpJ 2, and as such it is not fully backward compatible. This page shows how to migrate from CheerpJ 2 to CheerpJ 3 (or CheerpJ 4)
```

Either we just write (and above) to be forward compatible or we just write 4 explicitly for now

Author: We have "and above" everywhere at the moment.

> CheerpJ versions 3 and above are a complete reimplementation of CheerpJ 2, and as such, there is no backwards compatibility with CheerpJ 2. This page shows how to migrate from CheerpJ 2 to CheerpJ 3 and above, which is applicable for CheerpJ 4.

Reviewer: ```suggestion
CheerpJ version 3 (and above) is a complete reimplementation of CheerpJ 2, and as such it is not fully backward compatible. This page shows how to migrate from CheerpJ 2 to CheerpJ 3 and above, which is applicable for CheerpJ 4.
```

Most APIs are backward compatible, we should not overstate how many things are visibly changed.

### No Ahead-Of-Time compilation (`cheerpjify.py`)

> **No Ahead-Of-Time compilation:** To achieve good performance, CheerpJ required you to post-process JAR files with a custom binary compiler. The compiler would generate a `.jar.js` files for each input JAR. CheerpJ 3 (and versions above) features an advanced JIT engine that can generate better-performing code than CheerpJ 2.3 ever could. Removal of `.jar.js` files also significantly decreases how much data needs to be downloaded during application startup.

Reviewer: Either `and newer versions` or just `and above`

No downloads are provided with CheerpJ 3 and above.

- AOT optimization: CheerpJ 3 and above use a JIT compiler only, and as such it does not require any pre-processing of the JAR files, like conversion to `.jar.js`.
- `--natives`: JNI function implementations should be passed to `cheerpjInit` using the `natives` option. See the [JNI guide] for more information.

### Actual support for ClassLoaders

> CheerpJ 2.3 had very limited support for ClassLoaders. As a consequence of requiring AOT compilation of `.jar.js` files, it could only support the standard one provided by OpenJDK. CheerpJ 3 and above radically improves the status-quo by properly using ClassLoaders as expected by Java.

Reviewer: Proposal, should we settle for `3+`?

Author: I wonder if 3+ could create confusion of minor versions above 3.0 (e.g. 3.1, 3.2 etc but not CheerpJ 4)
